Man/imrot.1
.TH imrot(1) WCS "17 October 1996"
.SH Name
imrot \- Rotate and/or Reflect FITS and IRAF image files
.SH Synopsis
imrot [options] Fits or IRAF file(s)
.SH Description
.I imrot
is a utility for rotating and/or reflecting FITS or IRAF images.  Images can
be rotated only in multiples of 90 degrees.  The image may be output with
a different data type than that in which it is input, and IRAF files may be
written out as FITS files.
.SH Options
.B \-f
Write FITS image from IRAF input
.B \-l
Reflect image across vertical axis
.B \-o
Allow overwriting of input image, else write new one
.B \-r <angle>
Image rotation angle in degrees (default 0)
.B \-v
Verbose
.B \-x <num>
Output pixel size in bits (FITS code, default=input)
Allowed values are 8, 16, 32 for integers, -16 for unsigned integers, and
-32 and -64 for 4- and 8-byte floating point numbers
.SH Author
Doug Mink, SAO (dmink@cfa.harvard.edu)


NEWS
Version 3.6.7 (January 10, 2007)
bincat: Make catalog binning finally work right
sethead: Allow multiple keywords to be set from single line input file
wcs.h: Fix and add ANSI C prototypes
fitsfile.h, fitshead.h, wcscat.h: Add ANSI C prototypes
wcsinit.c: Change WCS letter argument from char to char* in wcsinitc(),wcsninitc()
findstar.c, imrotate.c: Include fitsfile.h instead of fitshead.h and imio.h
hget.c: Change WCS letter argument from char to char* in hget*c()
hget.c: Declare header and keyword const char in most subroutines
hput.c: Declare keyword and value const in most subroutines
imio.c: Include fitsfile.h instead of imio.h
imio.h: Drop because all declarations are in fitsfile.h
wcslib.h: Drop semicolon at end of c++ ifdef
wcslib.h: Drop second declaration of SZP subroutines
polfit.c: Move into catutil.c
Many smaller fixes were made so that standard C profiles work

Version 3.6.6 (November 6, 2006)
getdate: Add sidereal time conversions
imcat,scat: Fix SDSS returned data and ID number output
imwcs: Add -q b option to speed up star matching
remap: Check for output image off-scale as well as input image
remap: Fix bug setting scale
sethead: Do not read image unless header size in 2880-byte blocks is increased
sethead: Increase header buffer size when next keyword exceeds current length
skycoor: No longer print " degrees" if output in degrees
Allow right ascension and declination strings up to 31 characters everywhere
In wcsinit.c and wcscon.c, set projection to XY if units are pixels
ubcread.c: Fix header for nstarmax<1 output from ubcread()

Version 3.6.5 (July 5, 2006)
conpix: Add option to take base 10 log of entire image
edhead: Use mkstemp() instead of compiler-maligned tempnam()
scat: Print catalog object name if no number in single line search responses
imcat,immatch,imwcs,scat: Add support for SKYMAP 2000 catalog for wide fields
imcat,immatch,imwcs: Print catalog object name if no number
imsmooth,imresize: Write summary of operation to header
gethead: Fix bug which dropped trailing zeroes from strings
imrot,imsmooth: Deal with .ms.fits as a single extension
remap: Fix bug when adding to existing images
wcsinit: initialize 9x9 PC matrices; handle up to 9 axes for PC matrix
delwcs,delhead: Add -b option to leave blank lines in place of deleted keywords
hput.c: Fix alignment of comments in hputc() and hputcom(); add strfix()
ctgread.c,tabread.c: Add match argument to tabrnum() for sequential reads
Initialize uninitialized variables throughout package

Version 3.6.4 (May 8, 2006)
cphead: Fix bug so WCS headers copy correctly
gethead: Always underscore-pad not-tab null returns; make -a work correctly
gethead: Keep quoted strings as single value when reading free-format files
getcol: Count tokens on first line  with -k even if no columns are specified
imcat: Add second argument to -r to vary radii of PROS regions with magnitude
imcat,imwcs,immatch: Set image epoch from DATE-OBS or MJD-OBS, not EPOCH
scat,imcat,imwcs,immatch: Update SDSS access to DR4
scat,imcat: Add option to sort by ID number
scat: Fix bug so Hipparcos Catalog search returns magnitudes and parallax
scat: Clean up VOTable reporting
sethead: Fix _ substitution in COMMENT and HISTORY setting
imfill: New program to fill in bad pixels in an image; add IMFILL keyword
imsmooth: New program to smooth an image
imhead,imrot: Add +i to force appending of primary header keywords to FITS
extension headers.
imwcs,immatch: Set output image coordinate decimal places using nxydec=n
imhead, imsize, xy2sky, sky2xy, gethead: Read from fITS headers appended to
GIF, JPEG, or TIFF image files
sky2xy: Use -n number of decimal places when reading from file, too

binread.c: Fix bug so magnitudes are read when J2000 is set by negating the
           number of magnitudes
catutil.c:  Clean up VOTable code
catutil.c:  Return number of fields from vothead()
dateutil.c: Drop declarations of unused variables
fileutil.c: Add istiff(), isjpeg(), isgif() to check TIFF, JPEG, GIF files
filter.c:   Add subroutines to shrink an image
filter.c:   Add subroutines for Gaussian smoothing/filling
findstar.c: Add nxydec=num. decimal places in image coordinates to setparm()
            minpeak=num now sets minimum star counts above background
fitsfile.c: Add fitsrtail() to read FITS headers from end of non-FITS image
            files.  In fitsrhead(), add file name to all error messages
            Drop declarations of unused variables
imio.c:     Fix bug of occasional double application of bscale in getvec()
            Clean up arithmetic and increment in addpix() and multpix()
imsetwcs.c: Allow number of decimal places in image coordinates to be set
imutil.c: Add subroutines to filter, fill, or resize an image and reset its header
lwcs.h: Add NXYDEC to set default number of decimal places in image coordinates
sdssread.c: Change default URL's to search to SDSS DR4 site
sortstar.c: Add subroutines to sort by ID number
wcs.c       Drop declarations of unused variables
wcsinit.c:  Get Epoch of observation from MJD-OBS or DATE-OBS/UT unless DSS
            Set wcs->naxes to actual number of image WCS axes, usually 2
webread.c: Make changes so SDSS DR4 is processed correctly
           Multiply max stars for ESO search by fudge factor to get all;
           stars outside of box are returned (but not by scat in the end)

Version 3.6.3 (January 6, 2006)
scat: Fix bug reading sequence from catalog.  Clean up alignment of epoch
and velocity columns.  Search correctly with precessed center
scat, imcat: add 2MASS and Tycho-2 with magnitude errors.
imrot: Add option to flip image around horizontal axis
imrot: Add -p option to turn a negative image into a positive one
imstack: Fix bug which caused extra linefeeds
delhead: Fix bug which incorrectly padded with spaces when head is shrunk
getcol: Put line count on same line as mean or median values; add range of lines
wcscon.c: Fix bugs in rotation matrix computation for precession

Version 3.6.2 (July 29, 2005)

imrot.c
/* File imrot.c
 * February 24, 1998
 * By Doug Mink, Harvard-Smithsonian Center for Astrophysics
 * Send bug reports to dmink@cfa.harvard.edu
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <unistd.h>
#include <math.h>

#include "libwcs/fitshead.h"

static void usage();
static void imRot ();
extern int RotFITS();

static int verbose = 0;	/* verbose/debugging flag */
static int mirror = 0;	/* reflect image across vertical axis */
static int rotate = 0;	/* rotation in degrees, degrees counter-clockwise */
static int bitpix = 0;	/* number of bits per pixel (FITS code) */
static int fitsout = 0;	/* Output FITS file from IRAF input if 1 */
static int overwrite = 0;	/* allow overwriting of input image file */

main (ac, av)
int ac;
char **av;
{
    char *str;
    int readlist = 0;
    char *lastchar;
    char filename[128];
    FILE *flist;
    char *listfile;

    /* crack arguments */
    for (av++; --ac > 0 && (*(str = *av)=='-' || *str == '@'); av++) {
	char c;
	if (*str == '@')
	    str = str - 1;
	while (c = *++str)
    	switch (c) {
    	case 'f':	/* FITS file output */
    	    fitsout++;
    	    break;

    	case 'l':	/* image flipped around N-S axis */
	    mirror = 1;
    	    break;

	case 'o':	/* allow overwriting of existing image file */
	    overwrite++;
	    break;

    	case 'r':	/* Rotation angle in degrees */
    	    if (ac < 2)
    		usage ();
    	    rotate = (int) atof (*++av);
    	    ac--;
    	    break;

    	case 'x':	/* Number of bits per pixel */
    	    if (ac < 2)
    		usage ();
    	    bitpix = (int) atof (*++av);
    	    ac--;
    	    break;

    	case 'v':	/* more verbosity */
    	    verbose++;
    	    break;

	case '@':	/* List of files to be read */
	    readlist++;
	    listfile = ++str;
	    str = str + strlen (str) - 1;
	    av++;
	    ac--;

    	default:
    	    usage();
    	    break;
    	}
    }

    /* Find number of images to search and leave listfile open for reading */
    if (readlist) {
	if ((flist = fopen (listfile, "r")) == NULL) {
	    fprintf (stderr,"IMROT: List file %s cannot be read\n",
		     listfile);
	    usage ();
	    }
	while (fgets (filename, 128, flist) != NULL) {
	    lastchar = filename + strlen (filename) - 1;
	    if (*lastchar < 32) *lastchar = 0;
	    imRot (filename);
	    if (verbose)
		printf ("\n");
	    }
	fclose (flist);
	}

    /* If no arguments left, print usage */
    if (ac == 0)
	usage ();

    else {
	while (ac-- > 0) {
    	    char *fn = *av++;
    	    if (verbose)
    		printf ("%s:\n", fn);
    	    imRot (fn);
    	    if (verbose)
    		printf ("\n");
	    }
	}

    return (0);
}

static void
usage ()
{
    fprintf (stderr,"Rotate and/or Reflect FITS and IRAF image files\n");
    fprintf(stderr,"Usage: [-vm [-r rot] file.fts ...\n");
    fprintf(stderr,"  -f: write FITS image from IRAF input\n");
    fprintf(stderr,"  -l: reflect image across vertical axis\n");
    fprintf(stderr,"  -o: allow overwriting of input image, else write new one\n");
    fprintf(stderr,"  -r: image rotation angle in degrees (default 0)\n");
    fprintf(stderr,"  -v: verbose\n");
    fprintf(stderr,"  -x: output pixel size in bits (FITS code, default=input)\n");
    exit (1);
}

static void
imRot (name)
char *name;
{
    char *image;		/* FITS image */
    char *header;		/* FITS header */
    int lhead;			/* Maximum number of bytes in FITS header */
    int nbhead;			/* Actual number of bytes in FITS header */
    int iraffile;		/* 1 if IRAF image */
    int *irafheader;		/* IRAF image header */
    char newname[64];		/* Name for revised image */
    char *ext;
    char *imext, *imext1;
    char *fname;
    int lext, lroot;
    int bitpix0;
    char echar;
    char temp[8];
    char history[64];
    char pixname[128];

    /* If not overwriting input file, make up a name for the output file */
    if (!overwrite) {
	fname = strrchr (name, '/');
	if (fname)
	    fname = fname + 1;
	else
	    fname = name;
	ext = strrchr (fname, '.');
	if (ext != NULL) {
	    lext = (fname + strlen (fname)) - ext;
	    lroot = ext - fname;
	    strncpy (newname, fname, lroot);
	    *(newname + lroot) = 0;
	    }
	else {
	    lext = 0;
	    lroot = strlen (fname);
	    strcpy (newname, fname);
	    }
	imext = strchr (fname, ',');
	imext1 = NULL;
	if (imext == NULL) {
	    imext = strchr (fname, '[');
	    if (imext != NULL) {
		imext1 = strchr (fname, ']');
		*imext1 = (char) 0;
		}
	    }
	if (imext != NULL) {
	    strcat (newname, "_");
	    strcat (newname, imext+1);
	    }
	if (mirror)
	    strcat (newname, "m");
	else if (rotate != 0)
	    strcat (newname, "r");
	if (rotate < 10 && rotate > -1)
	    sprintf (temp,"%1d",rotate);
	else if (rotate < 100 && rotate > -10)
	    sprintf (temp,"%2d",rotate);
	else if (rotate < 1000 && rotate > -100)
	    sprintf (temp,"%3d",rotate);
	else
	    sprintf (temp,"%4d",rotate);
	if (rotate != 0)
	    strcat (newname, temp);
	if (fitsout)
	    strcat (newname, ".fit");
	else if (lext > 0) {
	    if (imext != NULL) {
		echar = *imext;
		*imext = (char) 0;
		strcat (newname, ext);
		*imext = echar;
		if (imext1 != NULL)
		    *imext1 = ']';
		}
	    else
		strcat (newname, ext);
	    }
	}
    else
	strcpy (newname, name);

    /* Open IRAF image if .imh extension is present */
    if (strsrch (name,".imh") != NULL) {
	iraffile = 1;
	if ((irafheader = irafrhead (name, &lhead)) != NULL) {
	    if ((header = iraf2fits (name, irafheader, lhead, &nbhead)) == NULL) {
		fprintf (stderr, "Cannot translate IRAF header %s/n",name);
		free (irafheader);
		return;
		}
	    if ((image = irafrimage (header)) == NULL) {
		hgets (header,"PIXFILE", 64, pixname);
		fprintf (stderr, "Cannot read IRAF pixel file %s\n", pixname);
		free (irafheader);
		free (header);
		return;
		}
	    }
	else {
	    fprintf (stderr, "Cannot read IRAF header file %s\n", name);
	    return;
	    }
	}

    /* Open FITS file if .imh extension is not present */
    else {
	iraffile = 0;
	if ((header = fitsrhead (name, &lhead, &nbhead)) != NULL) {
	    if ((image = fitsrimage (name, nbhead, header)) == NULL) {
		fprintf (stderr, "Cannot read FITS image %s\n", name);
		free (header);
		return;
		}
	    }
	else {
	    fprintf (stderr, "Cannot read FITS file %s\n", name);
	    return;
	    }
	}
    if (verbose) {
	fprintf (stderr,"Rotate and/or reflect ");
	if (iraffile)
	    fprintf (stderr,"IRAF image file %s", name);
	else
	    fprintf (stderr,"FITS image file %s", name);
	fprintf (stderr, " -> %s\n", newname);
	}

    if (bitpix != 0) {
	hgeti4 (header, "BITPIX", &bitpix0);
	if (verbose)
	    fprintf (stderr, "IMROT: %d bits/pixel -> %d bits/pixel\n",
		     bitpix0, bitpix);
	sprintf (history, "New copy of %s BITPIX %d -> %d",
		 name, bitpix0, bitpix);
	}
    else
	sprintf (history,"New copy of image %s", name);
    hputc (header,"HISTORY",history);

    if (RotFITS (name, header, &image, rotate, mirror, bitpix, verbose)) {
	fprintf (stderr,"Cannot rotate image %s; file is unchanged.\n",name);
	return;
	}
    if (bitpix != 0)
	hputi4 (header, "BITPIX", bitpix);
    if (iraffile && !fitsout) {
	if (irafwimage (newname, lhead, irafheader, header, image) > 0 && verbose)
	    printf ("%s: written successfully.\n", newname);
	}
    else {
	if (fitswimage (newname, header, image) > 0 && verbose)
	    printf ("%s: written successfully.\n", newname);
	}

    free (header);
    if (iraffile)
	free (irafheader);
    free (image);
    return;
}
/* Apr 15 1996	New program
 * Apr 18 1996	Add option to write to current working directory
 * May  2 1996	Pass filename to rotFITS
 * May 28 1996	Change rotFITS to RotFITS
 * Jun  6 1996	Always write to current working directory
 * Jun 14 1996	Use single image buffer
 * Jul  3 1996	Allow optional overwriting of input image
 * Jul 16 1996	Update header reading and allocation
 * Aug 26 1996	Change HGETC call to HGETS; pass LHEAD in IRAFWIMAGE
 * Aug 27 1996	Remove unused variables after lint
 *
 * Feb 21 1997  Check pointers against NULL explicitly for Linux
 * Aug 13 1997	Fix bug when overwriting an image
 * Dec 15 1997	Add capability of reading and writing IRAF 2.11 images
 *
 * Feb 24 1998	Add ext. to filename if writing part of multi-ext. file
 */
